### PR TITLE
Quantization(Dynamic and static)

### PR DESCRIPTION
**Pull request by CoderOMaster**
Hey @singjc need your suggestions on this notebook, i am relatively new to this topic but thanks for giving me the chance to work on it. I would like to ask you if its correct and is there anything incorrect which i need to sort out? Thanks please do provide feedback on this so i can improve it

### CODE REVIEW

**Comment by CoderOMaster** [comment]
one more thing i was trynna load these checkpoints and i got layer mismatch error which happened after quantisation so can we
 use this line in model_interface.py line 600 
 model.load_state_dict(checkpoint['model_state_dict'], strict=False) This will load the available keys and ignore any that are missing but would this be a good practice? and it resume at loss=0.0008 smthg then it crashed lol on cpu(why would that happen?) 
 
 If not then how should i proceed with static quantisation on checkpoint to preserve all layers? ig they would be better handled if i apply quantisation in main model files for qat then? what do u say ?

**Comment by CoderOMaster** [comment]
@singjc any update?

**Comment by singjc** [comment]
Sorry, I haven't had time to look at it fully. I'm pretty new to applying quantization as well, but we wanted to try it for this project, because in theory this generally should help reduce memory usage. So it's a learning process for us all :partying_face: 

Looking at the notebook and torch quantization documentation, I'm not sure if we want to pass the whole model directly to `quantization.prepare`? They use "Stubs" to mark where quantization should begin and end (converted back), I think we should probably use the same approach? We might also want to test which layers specifically would require quant vs which probably won't need it?

> one more thing i was trynna load these checkpoints and i got layer mismatch error which happened after quantisation so can we

Are the name/keys perhaps getting renamed after quantization? You could probably check the models state dict keys prior and post quantization to see which changes/is missing post quant

> model.load_state_dict(checkpoint['model_state_dict'], strict=False) This will load the available keys and ignore any that are missing but would this be a good practice?

Allows flexibility, but might cause unexpected crashes if the model tries to find a weight that doesn't exist. This would only be a problem if a different model state is used with a different version of the model though I guess.

> resume at loss=0.0008 smthg then it crashed lol on cpu(why would that happen?)

Is this the quantized model that crashes? Is there any error message that can help debug what the issue is? How does the system resources looks when resuming the training?

> If not then how should i proceed with static quantisation on checkpoint to preserve all layers? ig they would be better handled if i apply quantisation in main model files for qat then? what do u say ?

Yeah this is probably answered by my initial question/response above. I think we probably want to model the main model files to add optional quantization to the model structure itself.

@LLYX I don't know if you maybe already have experience with quantization / suggestions / comments

**Comment by CoderOMaster** [comment]
Okay I understand and we can use stubs in the architecture of model I guess(rest everything remains same i suppose bcoz we are passing relevant observers)? At the moment I was trying out how can I quantise existing checkpoint to load it on my computer and see if it works or not. But yeah ofc I am interested in applying quant and dequant to the main architecture too. And for keys mismatch I will manually print the one which are missing then I will quantise them bcoz it's hard to tell if they are getting renamed or not. Will print layers of both model checkpoint and fix that.

**Comment by CoderOMaster** [comment]
No error just my cpu usages increases a lot and it crashed the terminal. Will see what I can do.

**Comment by CoderOMaster** [comment]
Lol but it's funny that only that portion of terminal crashes rest everything intact

**Comment by CoderOMaster** [comment]
dict_keys(['epoch', 'model_state_dict', 'optimizer_state_dict', 'scheduler_state_dict', 'best_loss']) for BEST_MODEL.PTH(one u have)


dict_keys(['init_conv.weight', 'init_conv.bias', 'init_conv.scale', 'init_conv.zero_point', 'time_mlp.1.scale', 'time_mlp.1.zero_point', 'time_mlp.1._packed_params.dtype', 'time_mlp.1._packed_params._packed_params', 'time_mlp.3.scale', 'time_mlp.3.zero_point', ...... etc) for quantised model


Update: I have fixed this issue and issue was the dicr keys of quantised model were only part of model-state-dict and I added these along with best loss ,epoch and optimiser from original model directly to the quantised checkpoint. And I identified why it was crashing bcoz I had placed minimal quantization and size was only reduced to 10gb from 14gb. I will be back with more updates thanksm

**Comment by CoderOMaster** [comment]
@singjc what's your take on openvino ? Is it a good option to optimise the model?

Also even after adding quant dequang in unet I glt size of 2gb from 14gb while preserving relevant dicts but still it crashes( ram usage is 32gb aka it uses all the ram on system and all cores of cpu are above 90% usage) this should mean that it might be quantised and smaller in size but still weights are idk getting back to original precision somehow?

**Comment by CoderOMaster** [comment]
@singjc sorry i hv one more doubt that currently i am solely focusing on unet arch, should i keep on it or rather try on whole diffusion model?

Update: based on my research, quantising only unet architecture wont help much but rather its better to scale it up on diffusion model itself. I am reading on how to implement it better.

**Comment by singjc** [comment]
> @singjc what's your take on openvino ? Is it a good option to optimise the model?


I haven't personally used openvino myself, so can't really speak much on it. It does seem like a good option for Optimizing models though, but from my perusing of their docs, it seems like its more used for optimizing and deploying production ready models? I think we're still in the early development stage of figuring out the best model architecture to use.

> Also even after adding quant dequang in unet I glt size of 2gb from 14gb while preserving relevant dicts but still it crashes( ram usage is 32gb aka it uses all the ram on system and all cores of cpu are above 90% usage) this should mean that it might be quantised and smaller in size but still weights are idk getting back to original precision somehow?

I mean I think that's still already pretty good, shrinking down to 2Gb from 14Gb right. So quantization at least looks like the right approach to go down. Is it possible to print out the tensor data types of the loaded model and during training to see if it is being converted back to regular float precision?

> @singjc sorry i hv one more doubt that currently i am solely focusing on unet arch, should i keep on it or rather try on whole diffusion model?

Update: based on my research, quantising only unet architecture wont help much but rather its better to scale it up on diffusion model itself. I am reading on how to implement it better.

We don't need to stick to the Unet1 arch, it's just the only one that gave somewhat decent performance. We can explore other options. It's completely up to you if you want to spend more time optimizing the current unet or trying a different model.

**Comment by CoderOMaster** [comment]
Yeah I have had lot of stub experiments then I decided to read few papers on this. Currently I am reading this, it's a bit daunting but will do somehow ;)
https://github.com/Xiuyu-Li/q-diffusion/blob/master/scripts/sample_diffusion_ddim.py

**Comment by CoderOMaster** [comment]
My intuition is that weight are quantised but they behave like a compressed zip file analogy. Will look into it in depth

**Comment by CoderOMaster** [comment]
@singjc whats ur take on this ?
https://github.com/Xiuyu-Li/q-diffusion/issues/14
although its based on sd model but i have observed similar behaviour on ours

**Comment by singjc** [comment]
With the current setup, and with the old (JoshGPF) dataset, we pretty much use 100% of a single 40Gb A100 GPU. 

We can look at the GPU utilization for the most recent training experiment (20241116_lr_1eneg4_xo, 1.20B params)

image

image

image

image

image

**Comment by CoderOMaster** [comment]
Can you run it on cpu ? If if you are able to then can you show me the ram and cpu utilisation just before the model loads checkpoint bcoz I am using cpu now for the task( I no longer have access to a good gpu and currently I am working on 32gb Ram system with i7 9th)

**Comment by CoderOMaster** [comment]
Hi update is that I have abandoned dynamic quantization bcoz it works for limited torch.nn modules. Next js I had enough experiments with stubs in unet architecture but not much of help. Next I would try int4 in unet itself and as per papers I will be trying with quantization during timesteps. Let's see if I can make it work on cpu esp.

Update: Reason for ~3gb ckpt taking so much ram was that despite being compressed to int8 which is logically that 14gb orignal ckpt is at float32 so int8 one would be 14/4= 3.5 but they are not compatible with our custom layers and majority ones not compatible will load in fp32. So static quantization direct is bound to Fail.

**Comment by CoderOMaster** [comment]
Screenshot 2025-03-22 022409
Screenshot 2025-03-22 022417

size of ckpt is only reduced by 4gb, i was facing unstability with quantising bcoz in our custom model not all layers support quantisation so i settled in by making a wrapper class and initilising unet then passing quantised stub and then dequantising it.
If you think its okay then i can commit and continue working on more effective quantisation.

**Comment by CoderOMaster** [comment]
reason for less decrease in size is i cannot quantise everything on the unet as its fully supported by pytorch. one last thing(graph FX too)  i will try and see if its stable bcoz i faced a lot of instability issues. then if it works well i will commit that too. then move to the timestamp process quantisation and stuff which would be much better approach but this isnt bad either.

**Comment by CoderOMaster** [comment]
@singjc should i go for graph fx quantisation ? but for that i need to refactor our unet1d code as i get traceback error, have to remove if else statements, let me know what can we do for that?

**Comment by singjc** [comment]
> size of ckpt is only reduced by 4gb, i was facing unstability with quantising bcoz in our custom model not all layers support quantisation so i settled in by making a wrapper class and initilising unet then passing quantised stub and then dequantising it.
If you think its okay then i can commit and continue working on more effective quantisation.

I think this is fine, and a good start.

We could try graphFX, but if it's pretty unstable and would require a lot of effort, then we could maybe leave it for now?

**Comment by CoderOMaster** [comment]
Yeah it does look easy but it's not especially when you are using custom model with lots of if else. Not sure lol how will it perform even after so much effort. Can definitely try later. I will commit it then.

**Comment by CoderOMaster** [comment]
@singjc can you please check the nb and merge if its allright? thanks

**Comment by singjc** [comment]
> @singjc can you please check the nb and merge if its allright? thanks

Thanks for working on this! I just have a few suggestions/feedback.

- more of nitpick, can you move the nb to the `./nb` instead of having it in the main src code module.
- You have the whole model and the quantization wrapper in the nb cells, can you abstract that out into actual source code files in the main src code module (add doc-strings and doc to what you changed in the main model (if any) to make it work with the quantization.
- Can you also add some notes (markdown cells), with some information, explaining the purpose of the notebook, the steps you took, tools you used, and any explanations/discussion points. Just so that it's easier to understand what the purpose of the notebook is, what was done, what was evaluated, what is the end summary? Does it work? etc, Just so it makes it easier when we want to go back, to see what was done, if we need try something else out or just to recall what was done.

**Comment by CoderOMaster** [comment]
@singjc done, i have summarised main points of my research that would be helpful here.Would like to work more in depth by refering to papers i am interested in adapting.

**Comment by singjc** [comment]
Great, thanks! Will merge this for now.
